core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -391,6 +412,44 @@ void setupVersionMetrics() {
              .register(meterRegistry);
     }
 
+    /**
+     * Sets up gauge metric for each server certificate.
+     */
+    private void setupTlsMetrics(SslContext sslContext) {
+        final MeterRegistry meterRegistry = config().meterRegistry();
+
+        final SSLSession sslSession = validateSslContext(sslContext);
+        for (Certificate certificate : sslSession.getLocalCertificates()) {
+            try {
+                final X509Certificate x509Certificate = (X509Certificate) certificate;
+                final String commonName = getCommonName(x509Certificate);
+
+                final Supplier<Number> certificateExpirationInfoProvider = () -> {
+                    try {
+                        x509Certificate.checkValidity();
+                    } catch (CertificateExpiredException | CertificateNotYetValidException e) {
+                        return 1;
+                    }
+                    return 0;
+                };
+                Gauge.builder("armeria.server.certificate.expiration", certificateExpirationInfoProvider)
+                     .description("0 if certificate is not expired, 1 if certificate is expired")
+                     .tags("CN", commonName)
+                     .register(meterRegistry);
+
+                final Supplier<Number> timeToExpireProvider = () ->
+                        Duration.between(Instant.now(), x509Certificate.getNotAfter().toInstant()).getSeconds();
+                Gauge.builder("armeria.server.certificate.time.to.expire", timeToExpireProvider)
+                     .description("Duration in second before certificate expires which becomes negative " +
+                                  "if certificate is expired")
```

Comment:
- Precision
  Second precision seems too accurate for certification.
  Should we use days and return `-1` if the duration is negative?

- Meter naming
  JDK's `X509Certificate` uses `validity` as an official term. How about following the naming convention?
  https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/security/cert/X509Certificate.java#L153
  - `armeria.server.certificate.validity` could be 0 if certificate is expired.
  - `armeria.server.certificate.validity.days` to expose the remaining days?

Reply:
I agreed, will update as you suggested.